x-pack/plugins/cross_cluster_replication/server/plugin.ts

```diff
@@ -0,0 +1,135 @@
+import { PLUGIN } from '../common/constants';
+import { Dependencies } from './types';
+import { registerApiRoutes } from './routes';
+import { License } from './services';
+import { elasticsearchJsPlugin } from './client/elasticsearch_ccr';
+import { CrossClusterReplicationConfig } from './config';
+import { isEsError } from './lib/is_es_error';
+import { formatEsError } from './lib/format_es_error';
+
+interface CrossClusterReplicationContext {
+  client: IScopedClusterClient;
+}
+
+const ccrDataEnricher = async (indicesList: Index[], callWithRequest: APICaller) => {
+  if (!indicesList?.length) {
+    return indicesList;
+  }
+  const params = {
+    path: '/_all/_ccr/info',
+    method: 'GET',
+  };
+  try {
+    const { follower_indices: followerIndices } = await callWithRequest(
+      'transport.request',
+      params
+    );
+    return indicesList.map(index => {
+      const isFollowerIndex = !!followerIndices.find(
+        (followerIndex: { follower_index: string }) => {
+          return followerIndex.follower_index === index.name;
+        }
+      );
+      return {
+        ...index,
+        isFollowerIndex,
+      };
+    });
+  } catch (e) {
+    return indicesList;
+  }
+};
+
+export class CrossClusterReplicationServerPlugin implements Plugin<void, void, any, any> {
+  private readonly config$: Observable<CrossClusterReplicationConfig>;
+  private readonly license: License;
+  private readonly logger: Logger;
+
+  constructor(initializerContext: PluginInitializerContext) {
+    this.logger = initializerContext.logger.get();
+    this.config$ = initializerContext.config.create();
+    this.license = new License();
+  }
+
+  async setup(
+    { http, elasticsearch }: CoreSetup,
+    { licensing, indexManagement, remoteClusters }: Dependencies
+  ): Promise<void> {
+    const router = http.createRouter();
+    const config = await this.config$.pipe(first()).toPromise();
+
+    this.license.setup(
+      {
+        pluginId: PLUGIN.ID,
+        minimumLicenseType: PLUGIN.minimumLicenseType,
+        defaultErrorMessage: i18n.translate(
+          'xpack.crossClusterReplication.licenseCheckErrorMessage',
+          {
+            defaultMessage: 'License check failed',
+          }
+        ),
+      },
+      {
+        licensing,
+        logger: this.logger,
+      }
+    );
+
+    // Extend the elasticsearchJs client with additional endpoints.
+    const esClientConfig = { plugins: [elasticsearchJsPlugin] };
+    const ccrEsClient = elasticsearch.createClient('crossClusterReplication', esClientConfig);
+    http.registerRouteHandlerContext('crossClusterReplication', (ctx, request) => {
+      return {
+        client: ccrEsClient.asScoped(request),
+      };
+    });
+
+    registerApiRoutes({
+      router,
+      license: this.license,
+      lib: {
+        isEsError,
+        formatEsError,
+      },
+    });
+
+    // The UI is also dependent upon the Remote Clusters UI.
```

Comment:
The `remote_clusters.ui.enabled` config setting is what drives the value `remoteClusters.isUiEnabled`. It sounds like that's not clear? My intention for this comment was to explain that we care about this setting's value because the actual CCR UI cross-links to the Remote Clusters UI, so it won't function without it. I will add these details to the comment but let me know if there's anything else that isn't clear.